src/noteburst/jupyterclient/jupyterlab.py

```diff
@@ -0,0 +1,708 @@
+        status: int,
+        reason: Optional[str],
+        method: str,
+        body: Optional[str] = None,
+    ) -> None:
+        self.url = url
+        self.status = status
+        self.reason = reason
+        self.method = method
+        self.body = body
+        self.username = username
+        self.timestamp = datetime.datetime.now(tz=datetime.timezone.utc)
+        super().__init__(f"Status {status} from {method} {url}")
+
+    def __str__(self) -> str:
+        result = (
+            f"{self.username}: status {self.status} ({self.reason}) from"
+            f" {self.method} {self.url}"
+        )
+        if self.body:
+            result += f"\nBody:\n{self.body}\n"
+        return result
+
+
+class JupyterLabSessionError(Exception):
+    """An error related to a JupyterLab websocket session."""
+
+    def __init__(self, message: str, *, username: str) -> None:
+        self.username = username
+        super().__init__(message)
+
+    @classmethod
+    def from_exception(
+        cls, *, username: str, exception: Exception
+    ) -> JupyterLabSessionError:
+        return cls(str(exception), username=username)
+
+
+class CodeExecutionError(Exception):
+    """An error related to code execution in a JupyterLab session."""
+
+    def __init__(
+        self,
+        *,
+        username: str,
+        code: str,
+        code_type: str = "code",
+        error: Optional[str] = None,
+        status: Optional[str] = None,
+    ) -> None:
+        self.username = username
+        self.code = code
+        self.code_type = code_type
+        self.error = error
+        self.status = status
+        super().__init__("Code execution failed")
+
+    def __str__(self) -> str:
+        message = (
+            f"{self.username}: running {self.code_type} "
+            f"'{self.code}' failed"
+        )
+
+        message += f"\nError: {self.error}"
+        return message
+
+
+class JupyterClient:
+    """A client for JupyterLab, via JupyterHub.
+
+    This client should be used as a Python context. Each JupyterClient
+    includes its own HTTP client session to track cookies on behalf of the
+    user.
+
+    Parameters
+    ----------
+    user : `noteburst.jupyterclient.user.AuthenticatedUser`
+        The user's information.
+    logger : structlog.BoundLogger
+        A logger instance that may be associated with existing context.
+    config : JupyterConfig
+        Configuration for the JupyterLab image to spawn.
+    """
+
+    def __init__(
+        self,
+        *,
+        user: AuthenticatedUser,
+        logger: BoundLogger,
+        config: JupyterConfig,
+    ) -> None:
+        self.user = user
+        self.logger = logger
+        self.config = config
+
+        self.jupyter_url = urljoin(
+            noteburst_config.environment_url, self.config.url_prefix
+        )
+
+        self._http_client: Optional[httpx.AsyncClient] = None
+        self._cachemachine: Optional[CachemachineClient] = None
```

Comment:
This awkwardness and the need for tests and `RuntimeError` below are an indication that the type system isn't quite right. I think this all looks cleaner if, rather than having `JupyterClient` itself be the context manager, you add a (non-async) `connect` function that creates the http client and cachemachine client and then constructs a new `JupyterClientConnection` class which implements `__aenter__` and `__aexit__` and takes the `http_client` and `cachemachine` parameters as mandatory `__init__` parameters. Then `http_client` and `cachemachine` can be simple attributes of that object and the type system is always consistent.

Most of the methods would then be methods on `JupyterClientConnection` instead.